storage/storage.go

```diff
@@ -284,23 +284,37 @@ type bucketBoundHostname struct {
 	hostname string
 }
 
-func (s pathStyle) host(bucket string) string {
+func (s pathStyle) host(endpoint, bucket string) string {
+	if endpoint != "" {
+		return stripScheme(endpoint)
+	}
+
+	// This check is needed for clientless calls to SignedURL
 	if host := os.Getenv("STORAGE_EMULATOR_HOST"); host != "" {
 		return stripScheme(host)
 	}
 
+	// Fallback to default endpoint - clientless calls to SignURL/PostPolicy
+	// will not have an endpoint attached
 	return "storage.googleapis.com"
 }
 
-func (s virtualHostedStyle) host(bucket string) string {
+func (s virtualHostedStyle) host(endpoint, bucket string) string {
+	if endpoint != "" {
+		return bucket + "." + stripScheme(endpoint)
```

Comment:
I'm not totally sure about this one and whether it is expected to be supported. Let's discuss offline next week.